### DISCUSSION TITLE
How to run my own modules on pyodide

### DISCUSSION THREAD

**Discussion by LucasCelestinoSE** [pyodide/pyodide#4339] [Q&A] [answered]
Hello friend.
I have a problem that I can't solve.

I have this code with the filename display.py;

import hello

def main():
    hello.talk_hello()
main()

This is display dependence, called hello.py;

def talk_hello():
    print("Hello!")

Now the js code to run this program

try {
          let python_code = (await (await fetch('display.py')).text());
          await pyodide.runPythonAsync(python_code)
        } catch (err) {
          console.log(err);
        }

When I run pyodide, the following error appears:

PythonError: Traceback (most recent call last):
  File "/lib/python311.zip/_pyodide/_base.py", line 571, in eval_code_async
    await CodeRunner(
  File "/lib/python311.zip/_pyodide/_base.py", line 394, in run_async
    coroutine = eval(self.code, globals, locals)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<exec>", line 1, in <module>
ModuleNotFoundError: No module named 'hello'

I'm researching modules in pyodide, but I haven't found any way to run modules created by me, and I need to do this to continue with my project...

Thanks for helping me, and sorry if I did something that seemed beginner or dumb.

Thank you.

**Comment by ryanking13** [comment, marked as answer]
Pyodide uses a virtual file system, so you need to save your modules into the virtual file system to import them.

```js
// in JavaScript
const helloCode = await (await fetch('hello.py')).text()
pyodide.FS.writeFile("/home/pyodide/hello.py", helloCode)
pyodide.runPython("import hello; hello.do_something")
```

Note that you may need to set the `sys.path`  to make Python import system file your files.

```js
// By default Pyodide sets sys.path as follows
pyodide.runPython("import sys; print(sys.path)")
// ['/home/pyodide', '/lib/python311.zip', '/lib/python3.11', '/lib/python3.11/lib-dynload', '/lib/python3.11/site-packages']
```

Also, if your module has multiple files, it is not recommended to download each files separately as it will make bunch of requests, so you better make a package (.whl) or, make a zip archive and unpack it: see [FaQ](https://pyodide.org/en/stable/usage/faq.html#how-can-i-load-external-files-in-pyodide)

**Comment by LucasCelestinoSE** [comment]
Work for me, thank you for support ! <3
